### DISCUSSION TITLE
Let's talk about the CLA

### DISCUSSION THREAD

**Discussion by Zambito1** [jj-vcs/jj#3409] [General]
I would first like to thank all of the contributors to jj, and especially @martinvonz. Jj is seriously an awesome project. I use it every day both for work projects and personal projects. I appreciate the time that has been spent building this.

So everyone is on the same page, you can find a link to the CLA that is currently required to contribute to jj on this page: https://cla.developers.google.com/about

The CLA has been prohibitive to my contributing code to jj. I have opened several issues on this project, and through some dialog with other contributors I have been able to find the root cause of these issues. That is as far as I am willing to go with the CLA requirement. In the absense of the CLA, I would absolutely dig further and create pull requests.

I will not create a Google account to sign a license agreement (having a GitHub account is already a compromise to me). The license agreement doesn't do anything meaningful anyways. Google already has the right to distribute contributions under the permissive terms of the Apache license.

I recognize this post may not lead to any change in how contributions to the project are accepted, but I want this to be known instead of sitting on this information idly.

**Comment by martinvonz** [comment]
FWIW, I personally want Google to donate the project to a foundation, which would presumably mean that the Google CLA would go away. Again, that's just my personal opinion; I'm not speaking on Google's behalf, and I can't promise anything. Unless that happens, I think there's basically no chance of dropping the CLA.

> The license agreement doesn't do anything meaningful anyways.

IIUC what it does is to give Google (and others?) the option to release a fork under a different license in the future, without having to ask all past contributors for permission. But IANAL.

**Reply by Zambito1** [reply]
> I personally want Google to donate the project to a foundation, which would presumably mean that the Google CLA would go away.

That sounds like a great idea to me. I would also love to see that happen if possible.

>> The license agreement doesn't do anything meaningful anyways.

> IIUC what it does is to give Google (and others?) the option to release a fork under a different license in the future, without having to ask all past contributors for permission.

The reason I wrote that it doesn't do anything meaningful is because the Apache License specifically allows for this. See term 2:

>    2. Grant of Copyright License. Subject to the terms and conditions of
      this License, each Contributor hereby grants to You a perpetual,
      worldwide, non-exclusive, no-charge, royalty-free, irrevocable
      copyright license to reproduce, prepare Derivative Works of,
      publicly display, publicly perform, sublicense, and distribute the
      Work and such Derivative Works in Source or Object form.

This is essentially the same as the feature in the BSD license that enabled the recent Redis license change (The Redis organization used CLAs for other projects, but AFAIK Redis itself did not have a CLA. Either way the change was possible without one). Other people have taken advantage of the same ability to sublicense Redis to apply the LGPL lincese to it: https://codeberg.org/redict/redict/src/branch/main/LICENSES. Google, along with anyone else with access to the jj source code, can already apply an additional license to jj without asking any contributors. In fact, because the CLA does not include copyright assignment, Google does not have any particular extra ability related to changing the license; they cannot remove the Apache license, only apply additional terms under the sublicense clause.

Also obligitory IANAL, just a concerned citizen with an unusual interest in copyright licenses :P

**Reply by ilyagr** [reply]
As far as I understand, an important purpose of the (individual) CLA is that you assert that you have the right to your own work. That is, you affirm that Google doesn't need a corporate CLA from your employer to use your work.

> The reason I wrote that it doesn't do anything meaningful is because the Apache License specifically allows for this

For this reason, one could argue that signing the CLA doesn't cause any harm. Of course, it is an inconvenience, and IANAL and might be plain wrong.

**Reply by ilyagr** [reply]
I found out an interesting tidbit: Google's CLA [claims](https://opensource.google/documentation/reference/cla/#whats_in_the_cla) to be based on Apache's CLA, https://www.apache.org/licenses/contributor-agreements.html#clas, which is apparently required for any contribution to ASF project. One reason this is interesting to me is that ASF projects obviously use the Apache license.

**Comment by jrn** [comment]
Also discussed in https://github.com/jj-vcs/jj/discussions/4849.
